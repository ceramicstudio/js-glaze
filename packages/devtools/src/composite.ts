import type { CeramicApi } from '@ceramicnetwork/common'
import type {
  CompositeViewsDefinition,
  EncodedCompositeDefinition,
  InternalCompositeDefinition,
  Model,
  ModelDefinition,
  RuntimeCompositeDefinition,
  StreamCommits,
} from '@glazed/types'
import type { GraphQLSchema } from 'graphql'
import cloneDeep from 'lodash-es/cloneDeep.js'
import merge from 'lodash-es/merge.js'
import createObjectHash from 'object-hash'

import { decodeSignedMap, encodeSignedMap } from './formats/json.js'
import { createRuntimeDefinition } from './formats/runtime.js'
import { compositeModelsAndCommonEmbedsFromGraphQLSchema } from './schema.js'

type StrictCompositeDefinition = Required<InternalCompositeDefinition>

function toStrictDefinition(definition: InternalCompositeDefinition): StrictCompositeDefinition {
  return {
    aliases: {},
    commonEmbeds: [],
    views: { account: {}, root: {}, models: {} },
    ...definition,
  }
}

function isSupportedVersion(supported: string, check: string): boolean {
  const [supportedMajor] = supported.split('.')
  const [checkMajor] = check.split('.')
  return supportedMajor === checkMajor
}

function assertSupportedVersion(supported: string, check: string): void {
  if (!isSupportedVersion(supported, check)) {
    throw new Error(`Unsupported Composite version ${check}, expected version ${supported}`)
  }
}

function assertModelsHaveCommits(
  models: Record<string, ModelDefinition>,
  commits: Record<string, StreamCommits>
): void {
  for (const id of Object.keys(models)) {
    if (commits[id] == null) {
      throw new Error(`Missing commits for model ${id}`)
    }
  }
}

export function setDefinitionAliases(
  definition: StrictCompositeDefinition,
  aliases: Record<string, string>,
  replace = false
): StrictCompositeDefinition {
  const existing = replace ? {} : definition.aliases
  definition.aliases = { ...existing, ...aliases }
  return definition
}

export function setDefinitionCommonEmbeds(
  definition: StrictCompositeDefinition,
  names: Iterable<string>,
  replace = false
): StrictCompositeDefinition {
  const existing = replace ? [] : definition.commonEmbeds
  definition.commonEmbeds = Array.from(new Set([...existing, ...names]))
  return definition
}

export function setDefinitionViews(
  definition: StrictCompositeDefinition,
  views: CompositeViewsDefinition,
  replace = false
): StrictCompositeDefinition {
  const existing = replace ? {} : definition.views
  definition.views = merge(existing, views)
  return definition
}

async function loadModelsFromCommits<Models = Record<string, StreamCommits>>(
  _ceramic: CeramicApi,
  modelsCommits: Models
): Promise<Record<keyof Models, ModelDefinition>> {
  const definitions = await Promise.all(
    Object.values(modelsCommits).map((_commits): Promise<Model> => {
      throw new Error('Not implemented')
      // return await Model.fromCommits(ceramic, commits)
    })
  )
  return Object.keys(modelsCommits).reduce((acc, id, index) => {
    const model = definitions[index]
    if (model == null) {
      throw new Error(`Missing Model for id ${id}`)
    }
    const modelID = model.id.toString()
    if (modelID !== id) {
      throw new Error(`Unexpected Model stream ID: expected ${id}, got ${modelID}`)
    }
    acc[modelID as keyof Models] = model.content
    return acc
  }, {} as Record<keyof Models, ModelDefinition>)
}

export type CompositeParams = {
  commits: Record<string, StreamCommits>
  definition: InternalCompositeDefinition
}

export type CompositeInput = Composite | CompositeParams
export type CompositeOptions = {
  aliases?: Record<string, string>
  commonEmbeds?: 'all' | 'none' | Iterable<string>
  views?: CompositeViewsDefinition
}

export type CreateParams = {
  ceramic: CeramicApi
  schema: string | GraphQLSchema
}

export type FromJSONParams = {
  ceramic: CeramicApi
  definition: EncodedCompositeDefinition
}

export type FromModelsParams = CompositeOptions & {
  ceramic: CeramicApi
  models: Array<string>
}

export class Composite {
  static VERSION = '1.0'

  static create(params: CreateParams): Promise<Composite> {
    // TODO: convert the schema to models, create the models on the Ceramice node, load their commits

    const { models, commonEmbeds } = compositeModelsAndCommonEmbedsFromGraphQLSchema(params.schema)
    console.log('MODELS', models)
    console.log('COMMON EMBEDS', commonEmbeds)

    throw new Error('Not implemented')
  }

  static from(composites: Iterable<CompositeInput>, options?: CompositeOptions): Composite {
    const [first, ...rest] = composites
    if (first == null) {
      throw new Error('Missing composites to compose')
    }
    const composite = first instanceof Composite ? first.clone() : new Composite(first)
    return composite.merge(rest, options)
  }

  static async fromJSON(params: FromJSONParams): Promise<Composite> {
    const { models, ...definition } = params.definition
    const commits = decodeSignedMap(models)
    return new Composite({
      commits,
      definition: {
        ...definition,
        models: await loadModelsFromCommits(params.ceramic, commits),
      },
    })
  }

  static fromModels(_params: FromModelsParams): Promise<Composite> {
    // TODO: load all models definitions and commits
    throw new Error('Not implemented')
  }

  #commits: Record<string, StreamCommits>
  #definition: StrictCompositeDefinition
  #hash: string | undefined

  constructor(params: CompositeParams) {
    assertSupportedVersion(Composite.VERSION, params.definition.version)
    assertModelsHaveCommits(params.definition.models, params.commits)
    this.#commits = cloneDeep(params.commits)
    this.#definition = toStrictDefinition(cloneDeep(params.definition))
  }

  get hash(): string {
    if (this.#hash == null) {
      this.#hash = createObjectHash(this.#definition)
    }
    return this.#hash
  }

  clone(): Composite {
    return new Composite(this.toParams())
  }

  copy(models: Array<string>): Composite {
    if (models.length === 0) {
      throw new Error('Missing models to copy')
    }

    const { commits, definition } = this.toParams()
    const def = toStrictDefinition(definition)

    const nameIDs = Object.entries(def.models).reduce((acc, [id, model]) => {
      acc[model.name] = id
      return acc
    }, {} as Record<string, string>)
    const aliasIDs = Object.entries(def.aliases).reduce((acc, [id, alias]) => {
      acc[alias] = id
      return acc
    }, {} as Record<string, string>)

    const nextCommits: Record<string, StreamCommits> = {}
    const nextModels: Record<string, ModelDefinition> = {}
    const nextAliases: Record<string, string> = {}
    const nextViews: CompositeViewsDefinition = { account: {}, root: {}, models: {} }

    for (const model of models) {
      const id = aliasIDs[model] ?? nameIDs[model] ?? model
      if (def.models[id] == null) {
        throw new Error(`Model not found: ${model}`)
      }

      nextCommits[id] = commits[id]
      nextModels[id] = def.models[id]
      if (def.aliases[id] != null) {
        nextAliases[id] = def.aliases[id]
      }
      if (def.views.models[id] != null) {
        // TODO: check relations to other models, ensure there's no missing model in the subset
        nextViews.models[id] = def.views.models[id]
      }
      // TODO: account and root views
    }

    return new Composite({
      commits: nextCommits,
      definition: {
        version: def.version,
        commonEmbeds: def.commonEmbeds,
        models: nextModels,
        aliases: nextAliases,
        views: nextViews,
      },
    })
  }

  equals(other: CompositeInput): boolean {
    const otherHash =
      other instanceof Composite
        ? other.hash
        : createObjectHash(toStrictDefinition(other.definition))
    return this.hash === otherHash
  }

  merge(other: CompositeInput | Array<CompositeInput>, options: CompositeOptions = {}): Composite {
    const nextParams = this.toParams()
    const nextDefinition = toStrictDefinition(nextParams.definition)
    const collectedEmbeds = new Set<string>()

    for (const composite of Array.isArray(other) ? other : [other]) {
      const { commits, definition } =
        composite instanceof Composite ? composite.toParams() : composite
      assertSupportedVersion(nextDefinition.version, definition.version)
      assertModelsHaveCommits(definition.models, commits)

      const def = toStrictDefinition(definition)
      Object.assign(nextParams.commits, commits)
      Object.assign(nextDefinition.models, definition.models)
      Object.assign(nextDefinition.aliases, def.aliases)
      merge(nextDefinition.views, def.views)
      for (const name of def.commonEmbeds) {
        collectedEmbeds.add(name)
      }
    }

    if (options.aliases != null) {
      setDefinitionAliases(nextDefinition, options.aliases)
    }
    const commonEmbeds = options.commonEmbeds ?? 'none'
    if (commonEmbeds === 'all') {
      setDefinitionCommonEmbeds(nextDefinition, collectedEmbeds)
    } else if (Array.isArray(commonEmbeds)) {
      setDefinitionCommonEmbeds(nextDefinition, commonEmbeds, true)
    }
    if (options.views != null) {
      setDefinitionViews(nextDefinition, options.views)
    }

    return new Composite({ ...nextParams, definition: nextDefinition })
  }

  setAliases(aliases: Record<string, string>, replace = false): Composite {
    const params = this.toParams()
    const definition = setDefinitionAliases(toStrictDefinition(params.definition), aliases, replace)
    return new Composite({ ...params, definition })
  }

  setCommonEmbeds(names: Iterable<string>, replace = false): Composite {
    const params = this.toParams()
    const definition = setDefinitionCommonEmbeds(
      toStrictDefinition(params.definition),
      names,
      replace
    )
    return new Composite({ ...params, definition })
  }

  setViews(views: CompositeViewsDefinition, replace = false): Composite {
    const params = this.toParams()
    const definition = setDefinitionViews(toStrictDefinition(params.definition), views, replace)
    return new Composite({ ...params, definition })
  }

  toJSON(): EncodedCompositeDefinition {
    return {
      version: this.#definition.version,
      models: encodeSignedMap(this.#commits),
      aliases: this.#definition.aliases,
      views: this.#definition.views,
    }
  }

  toParams(): CompositeParams {
    return {
      commits: cloneDeep(this.#commits),
      definition: cloneDeep(this.#definition),
    }
  }

  toRuntime(): RuntimeCompositeDefinition {
    return createRuntimeDefinition(this.#definition)
  }
}
