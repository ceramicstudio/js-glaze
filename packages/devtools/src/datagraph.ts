export class DataGraph {
  // Deterministic account to single record (IDX)
  async getRecord(type: string, did?: string): Promise<any> {}
  async setRecord(type: string, content: any): Promise<any> {}
  async removeRecord(type: string): Promise<any> {}

  // Account to multiple records
  async listCollectionRecords(type: string, did?: string): Promise<any> {}
  async addCollectionRecord(type: string, content: any): Promise<any> {}
  async removeCollectionRecord(id: string): Promise<any> {}

  // Deterministic account links
  async getAccountLink(type: string, did: string): Promise<any> {}
  async setAccountLink(type: string, did: string, content?: any): Promise<any> {}
  async removeAccountLink(type: string, did: string): Promise<any> {}

  // Account links query
  async listAccountLinksFrom(type: string, did?: string): Promise<any> {}
  async listAccountLinksTo(type: string, did?: string): Promise<any> {}
  async listReciprocalAccountLinks(type: string, did?: string): Promise<any> {}
  async listAccountLinksRecords(
    linkType: string,
    collectionType: string,
    did?: string
  ): Promise<any> {}

  // Records associated to stream -- should these use the CollectionRecords APIs?
  // Need to define query parameters for collections
  async listChildrenRecords(streamID: string, type: string): Promise<any> {}
  async addChildRecord(streamID: string, type: string, content: any): Promise<any> {}
  // Should this simply be removeCollectionRecord?
  async removeChildRecord(streamID: string, type: string, id: string): Promise<any> {}

  // TODO: getCollection with all basic profiles
}
