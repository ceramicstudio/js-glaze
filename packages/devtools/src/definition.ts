export type DefinitionNodeType =
  | 'self' // default
  | 'did'
  | 'streamid'
  | 'any'

export type DefinitionNodeConfig = {
  type: DefinitionNodeType
  optional?: boolean // defaults to `false`
}

export type DefinitionConfig = {
  multiple?: boolean // defaults to `false` = deterministic stream (IDX default)
  to?: DefinitionNodeConfig
  from?: DefinitionNodeConfig
}

function isSelfNodeConfig(config?: DefinitionNodeConfig): boolean {
  return config == null || config.type === 'self'
}

// definition queries/mutations based on config type:
// - if multiple == false or not set, and to and from are "self" or unset -> IDX deterministic get/set
// - if multiple == false or not set

export function getDefinitionAccessPattern(config: DefinitionConfig = {}) {
  const isFromSelf = isSelfNodeConfig(config.from)
  const isToSelf = isSelfNodeConfig(config.to)

  if (config.multiple) {
    if (isFromSelf) {
      if (isToSelf) {
        // Collection list/add/remove with content on stream, ex: posts
      } else {
        // Collection list/add/remove linkTo, edge content based on schema, ex: comments on post
        // Could be optional link - in that case similar to previous logic
      }
    } else if (isToSelf) {
      // Collection of links with specified subject but object is stream content - what use case?
    } else {
      // Collection of links with specified subject and object = "any edge" use-case
    }
  } else if (isFromSelf) {
    if (isToSelf) {
      // Deterministic get/set/remove content == IDX
    } else {
      // Deterministic get/set/remove linkTo, edge content based on schema, ex: account link
    }
  } else if (isToSelf) {
    // Deterministic link with specified subject but object is stream content - what use case?
  } else {
    // Deterministic link with specified subject and object - what use case?
  }
}

type PosterModel = {
  v: 1
  protocols: {
    schemas: {
      BasicProfile: {}
      Post: {}
      EmptySchema: {}
    }
    definitions: {
      basicProfile: {
        // read/write: deterministic/IDX
        schema: '<BasicProfile>'
      }
      post: {
        // read: connection with optional streamID filter (false | string)
        // write: add to connection, optional linkTo param
        schema: '<Post>'
        config: {
          multiple: true
          to: { type: 'streamid'; optional: true } // Can be a reply or original post
        }
      }
      follow: {
        // read: based on DID type, connection from + connection to
        // write: add/remove follow link
        schema: '<EmptySchema>'
        config: {
          multiple: false
          to: { type: 'did' }
        }
      }
    }
  }
}
