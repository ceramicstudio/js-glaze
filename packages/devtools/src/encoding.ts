import type { EncodedDagJWS, EncodedDagJWSResult, ManagedEntry, ManagedModel } from '@glazed/types'
import CID from 'cids'
import type { DagJWS, DagJWSResult } from 'dids'
import { fromString, toString } from 'uint8arrays'

import { applyMap } from './utils'

export function decodeDagJWS({ payload, signatures, link }: EncodedDagJWS): DagJWS {
  return { payload, signatures, link: link ? new CID(link) : undefined }
}

export function encodeDagJWS({ payload, signatures, link }: DagJWS): EncodedDagJWS {
  return { payload, signatures, link: link?.toString() }
}

export function decodeDagJWSResult({ jws, linkedBlock }: EncodedDagJWSResult): DagJWSResult {
  return { jws: decodeDagJWS(jws), linkedBlock: fromString(linkedBlock, 'base64pad') }
}

export function encodeDagJWSResult({ jws, linkedBlock }: DagJWSResult): EncodedDagJWSResult {
  return { jws: encodeDagJWS(jws), linkedBlock: toString(linkedBlock, 'base64pad') }
}

/** @internal */
export function decodeSignedMap<K extends string>(
  data: Record<K, Array<EncodedDagJWSResult>>
): Record<K, Array<DagJWSResult>> {
  return applyMap(data, (records) => records.map(decodeDagJWSResult))
}

/** @internal */
export function encodeSignedMap<K extends string>(
  data: Record<K, Array<DagJWSResult>>
): Record<K, Array<EncodedDagJWSResult>> {
  return applyMap(data, (records) => records.map(encodeDagJWSResult))
}

/** @internal */
export function decodeEntryCommits(
  entry: ManagedEntry<EncodedDagJWSResult>
): ManagedEntry<DagJWSResult> {
  return { ...entry, commits: entry.commits.map(decodeDagJWSResult) }
}

export function decodeModel(model: ManagedModel<EncodedDagJWSResult>): ManagedModel<DagJWSResult> {
  return {
    schemas: applyMap(model.schemas, (schema) => {
      return { ...schema, commits: schema.commits.map(decodeDagJWSResult) }
    }),
    definitions: applyMap(model.definitions, decodeEntryCommits),
    tiles: applyMap(model.tiles, decodeEntryCommits),
  }
}

/** @internal */
export function encodeEntryCommits(
  entry: ManagedEntry<DagJWSResult>
): ManagedEntry<EncodedDagJWSResult> {
  return { ...entry, commits: entry.commits.map(encodeDagJWSResult) }
}

export function encodeModel(model: ManagedModel<DagJWSResult>): ManagedModel<EncodedDagJWSResult> {
  return {
    schemas: applyMap(model.schemas, (schema) => {
      return { ...schema, commits: schema.commits.map(encodeDagJWSResult) }
    }),
    definitions: applyMap(model.definitions, encodeEntryCommits),
    tiles: applyMap(model.tiles, encodeEntryCommits),
  }
}
