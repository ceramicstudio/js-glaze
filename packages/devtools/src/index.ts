/**
 * ```sh
 * npm install --dev @glazed/devtools
 * ```
 *
 * @module devtools
 */

export * from './datamodel'
export * from './encoding'
export * from './graphql'
export * from './publishing'
export * from './schema'
export * from './utils'
export * from './validation'
