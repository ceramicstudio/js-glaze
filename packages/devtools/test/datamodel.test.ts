/**
 * @jest-environment glaze
 */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */

import { ModelManager, publishEncodedModel } from '../src'

describe('datamodel', () => {
  jest.setTimeout(20000)

  test('publish encoded model', async () => {
    const encodedModel = {
      schemas: {
        kjzl6cwe1jw14amy1imkbql1d61u00q9cbvhy5c3jtv3nz552fshl013530rauh: {
          alias: 'DataStoreIndex',
          commits: [
            {
              jws: {
                payload: 'AXESIKO29HNjXsFTO9Bbs5VTTGqJUPZ9Aoic83L6G0ziXp48',
                signatures: [
                  {
                    signature:
                      'R-VH5EwcOsvd3Txsl4Pjw-QSobjiNJuQVPWrbubyX6uIbe2SRupKx5od0z2agdeVnglt9L9mHiOqlRGgI7WNDQ',
                    protected:
                      'eyJhbGciOiJFZERTQSIsImtpZCI6ImRpZDprZXk6ejZNa3NOWUU2V3RNWjNXTGJQd2NwOUdtN2RUN0ozRHNOUDhhdVFOblBwUUJ1MzdBI3o2TWtzTllFNld0TVozV0xiUHdjcDlHbTdkVDdKM0RzTlA4YXVRTm5QcFFCdTM3QSJ9',
                  },
                ],
                link: 'bafyreifdw32hgy26yfjtxuc3wokvgtdkrfipm7icrcopg4x2dngoexu6hq',
              },
              linkedBlock:
                'o2RkYXRhpWR0eXBlZm9iamVjdGV0aXRsZW1JZGVudGl0eUluZGV4ZyRzY2hlbWF4J2h0dHA6Ly9qc29uLXNjaGVtYS5vcmcvZHJhZnQtMDcvc2NoZW1hI2tkZWZpbml0aW9uc6FsQ2VyYW1pY0RvY0lko2R0eXBlZnN0cmluZ2dwYXR0ZXJueBxeY2VyYW1pYzovLy4rKFw/dmVyc2lvbj0uKyk/aW1heExlbmd0aBiWdGFkZGl0aW9uYWxQcm9wZXJ0aWVzoWQkcmVmeBojL2RlZmluaXRpb25zL0NlcmFtaWNEb2NJZGZoZWFkZXKiZnNjaGVtYfdrY29udHJvbGxlcnOBeDhkaWQ6a2V5Ono2TWtzTllFNld0TVozV0xiUHdjcDlHbTdkVDdKM0RzTlA4YXVRTm5QcFFCdTM3QWdkb2N0eXBlZHRpbGU=',
            },
          ],
          dependencies: {},
          version: 'k3y52l7qbv1fryjn62sggjh1lpn11c56qfofzmty190d62hwk1cal1c7qc5he54ow',
        },
        kjzl6cwe1jw1482rpzfuczmbqkxnevw3risxar23d7z2majhkm9pouujiov58tq: {
          alias: 'DataStoreDefinition',
          commits: [
            {
              jws: {
                payload: 'AXESIGVBGu8EHXq5fxC16J4MwEPBCfrWq1bsZWQP2F_WWhT0',
                signatures: [
                  {
                    signature:
                      'hvv_KWbyYbg7ZP8PQeJxPjHnlAg1VKtPjVa6zXwMq9pug96qGcXFQi5XUAl30RH4QYOHl8Gpv4uZczWxAJllAg',
                    protected:
                      'eyJhbGciOiJFZERTQSIsImtpZCI6ImRpZDprZXk6ejZNa3NOWUU2V3RNWjNXTGJQd2NwOUdtN2RUN0ozRHNOUDhhdVFOblBwUUJ1MzdBI3o2TWtzTllFNld0TVozV0xiUHdjcDlHbTdkVDdKM0RzTlA4YXVRTm5QcFFCdTM3QSJ9',
                  },
                ],
                link: 'bafyreidfieno6ba5pk4x6efv5cpazqcdyee7vvvlk3wgkzap3bp5mwqu6q',
              },
              linkedBlock:
                'o2RkYXRhpmR0eXBlZm9iamVjdGV0aXRsZWpEZWZpbml0aW9uZyRzY2hlbWF4J2h0dHA6Ly9qc29uLXNjaGVtYS5vcmcvZHJhZnQtMDcvc2NoZW1hI2hyZXF1aXJlZINkbmFtZWtkZXNjcmlwdGlvbmZzY2hlbWFqcHJvcGVydGllc6VjdXJsomR0eXBlZnN0cmluZ2ltYXhMZW5ndGgY8GRuYW1lomR0eXBlZnN0cmluZ2ltYXhMZW5ndGgYlmZjb25maWehZHR5cGVmb2JqZWN0ZnNjaGVtYaFkJHJlZngaIy9kZWZpbml0aW9ucy9DZXJhbWljRG9jSWRrZGVzY3JpcHRpb26iZHR5cGVmc3RyaW5naW1heExlbmd0aBkBpGtkZWZpbml0aW9uc6FsQ2VyYW1pY0RvY0lko2R0eXBlZnN0cmluZ2dwYXR0ZXJueBxeY2VyYW1pYzovLy4rKFw/dmVyc2lvbj0uKyk/aW1heExlbmd0aBiWZmhlYWRlcqJmc2NoZW1h92tjb250cm9sbGVyc4F4OGRpZDprZXk6ejZNa3NOWUU2V3RNWjNXTGJQd2NwOUdtN2RUN0ozRHNOUDhhdVFOblBwUUJ1MzdBZ2RvY3R5cGVkdGlsZQ==',
            },
          ],
          dependencies: {},
          version: 'k3y52l7qbv1fry1fp4s0nwdarh0vahusarpposgevy0pemiykymd2ord6swtharcw',
        },
        kjzl6cwe1jw146mi4smwraxjypxq5d1qwq9iei4yydyxf6jwna1d8wouohto87e: {
          alias: 'Note',
          commits: [
            {
              jws: {
                payload: 'AXESIKiye0AVOuPrc8DMbos2QCqi4AZovd6PfkRRry7xqynW',
                signatures: [
                  {
                    signature:
                      'fjs2dgtTkJ59rHxIigfdnGPEVsDn_tFjT0uM_05vgUcs6_UsZVH7GZff3ylLT21Crd6qG_FqLPCfuhCzblyHDA',
                    protected:
                      'eyJhbGciOiJFZERTQSIsImtpZCI6ImRpZDprZXk6ejZNa2lUQnoxeW11ZXBBUTRIRUhZU0YxSDhxdUc1R0xWVlFSM2RqZFgzbURvb1dwI3o2TWtpVEJ6MXltdWVwQVE0SEVIWVNGMUg4cXVHNUdMVlZRUjNkamRYM21Eb29XcCJ9',
                  },
                ],
                link: 'bafyreifiwj5uafj24pvxhqgmn2ftmqbkulqam2f532hx4rcrv4xpdkzj2y',
              },
              linkedBlock:
                'omRkYXRhpWR0eXBlZm9iamVjdGV0aXRsZWROb3RlZyRzY2hlbWF4J2h0dHA6Ly9qc29uLXNjaGVtYS5vcmcvZHJhZnQtMDcvc2NoZW1hI2hyZXF1aXJlZIBqcHJvcGVydGllc6JkZGF0ZaRkdHlwZWZzdHJpbmdldGl0bGVkZGF0ZWZmb3JtYXRpZGF0ZS10aW1laW1heExlbmd0aBgeZHRleHSjZHR5cGVmc3RyaW5nZXRpdGxlZHRleHRpbWF4TGVuZ3RoGQ+gZmhlYWRlcqJmdW5pcXVlcGNWazlHSkIxejZzU3JYOHJrY29udHJvbGxlcnOBeDhkaWQ6a2V5Ono2TWtpVEJ6MXltdWVwQVE0SEVIWVNGMUg4cXVHNUdMVlZRUjNkamRYM21Eb29XcA==',
            },
          ],
          dependencies: {},
          version: 'k3y52l7qbv1frxr40y3mwy5qlywfty4sfcqrmv4zckjcdnylk5jdq7cgq66r0acjk',
        },
        kjzl6cwe1jw149knipvwi8blih9n07be1f82wlps0qe4pjgq8qlpjgm4ppgi87v: {
          alias: 'NotesList',
          commits: [
            {
              jws: {
                payload: 'AXESIFk6k_JCPGisehqBXwSKS-15EqsrM_q6fv_BgzBUQqFg',
                signatures: [
                  {
                    signature:
                      'GtCSPTsh-Q2aF-cdDegEseH59fIgOT10DnZyzoG6usbxq41cHGTQGK4_pfwS1W97xntjbQ-IU8pC7J6nd4loDQ',
                    protected:
                      'eyJhbGciOiJFZERTQSIsImtpZCI6ImRpZDprZXk6ejZNa2lUQnoxeW11ZXBBUTRIRUhZU0YxSDhxdUc1R0xWVlFSM2RqZFgzbURvb1dwI3o2TWtpVEJ6MXltdWVwQVE0SEVIWVNGMUg4cXVHNUdMVlZRUjNkamRYM21Eb29XcCJ9',
                  },
                ],
                link: 'bafyreiczhkj7eqr4ncwhugubl4cius7npejkwkzt7k5h576bqmyfiqvbma',
              },
              linkedBlock:
                'omRkYXRhpWR0eXBlZm9iamVjdGV0aXRsZWlOb3Rlc0xpc3RnJHNjaGVtYXgnaHR0cDovL2pzb24tc2NoZW1hLm9yZy9kcmFmdC0wNy9zY2hlbWEjaHJlcXVpcmVkgGpwcm9wZXJ0aWVzoWVub3Rlc6NkdHlwZWVhcnJheWVpdGVtc6RkdHlwZWZvYmplY3RldGl0bGVoTm90ZUl0ZW1ocmVxdWlyZWSAanByb3BlcnRpZXOiYmlkomR0eXBlZnN0cmluZ2gkY29tbWVudHhVY2lwODg6cmVmOmNlcmFtaWM6Ly9rM3k1Mmw3cWJ2MWZyeHI0MHkzbXd5NXFseXdmdHk0c2ZjcXJtdjR6Y2tqY2RueWxrNWpkcTdjZ3E2NnIwYWNqa2V0aXRsZaNkdHlwZWZzdHJpbmdldGl0bGVldGl0bGVpbWF4TGVuZ3RoGGRldGl0bGVlbm90ZXNmaGVhZGVyomZ1bmlxdWVwWGNPR3U1N2R6b08wZVd4Smtjb250cm9sbGVyc4F4OGRpZDprZXk6ejZNa2lUQnoxeW11ZXBBUTRIRUhZU0YxSDhxdUc1R0xWVlFSM2RqZFgzbURvb1dw',
            },
          ],
          dependencies: {
            'notes.id': ['kjzl6cwe1jw146mi4smwraxjypxq5d1qwq9iei4yydyxf6jwna1d8wouohto87e'],
          },
          version: 'k3y52l7qbv1fryc2v942v5n6gzesjlg0zu49gnufb97ohhmeym56dme5dit1dmfwg',
        },
      },
      definitions: {
        kjzl6cwe1jw145d00g4tuhpps6ozv30hb30cnlaevxmhy8hqbo2x8w58bupuhua: {
          alias: 'myNotes',
          commits: [
            {
              jws: {
                payload: 'AXESICreMVjrLh2HUVYljVDC1EXuQCwkx8sh47ndOmpaLY5K',
                signatures: [
                  {
                    signature:
                      '6gHBHJrAsmbQipRtXR0yQXhssKIuiHG2TisR443NkIsXqtbMrFu7IodxGhJAz_SdZYimgfomqgOr_Hrq2IKJBg',
                    protected:
                      'eyJhbGciOiJFZERTQSIsImtpZCI6ImRpZDprZXk6ejZNa2lUQnoxeW11ZXBBUTRIRUhZU0YxSDhxdUc1R0xWVlFSM2RqZFgzbURvb1dwI3o2TWtpVEJ6MXltdWVwQVE0SEVIWVNGMUg4cXVHNUdMVlZRUjNkamRYM21Eb29XcCJ9',
                  },
                ],
                link: 'bafyreibk3yyvr2zodwdvcvrfrvimfvcf5zacyjghzmq6hoo5hjvfulmoji',
              },
              linkedBlock:
                'omRkYXRho2RuYW1lZW5vdGVzZnNjaGVtYXhLY2VyYW1pYzovL2szeTUybDdxYnYxZnJ5YzJ2OTQydjVuNmd6ZXNqbGcwenU0OWdudWZiOTdvaGhtZXltNTZkbWU1ZGl0MWRtZndna2Rlc2NyaXB0aW9uaE15IG5vdGVzZmhlYWRlcqNmc2NoZW1heEtjZXJhbWljOi8vazN5NTJsN3FidjFmcnkxZnA0czBud2RhcmgwdmFodXNhcnBwb3NnZXZ5MHBlbWl5a3ltZDJvcmQ2c3d0aGFyY3dmdW5pcXVlcFZmSUd1SkIrYkNFbjJEL25rY29udHJvbGxlcnOBeDhkaWQ6a2V5Ono2TWtpVEJ6MXltdWVwQVE0SEVIWVNGMUg4cXVHNUdMVlZRUjNkamRYM21Eb29XcA==',
            },
          ],
          schema: 'kjzl6cwe1jw149knipvwi8blih9n07be1f82wlps0qe4pjgq8qlpjgm4ppgi87v',
          version: 'k3y52l7qbv1frxi4g36qc8tyvcflr11ff2tehzre1v33znge36z8sf8l78afsuvb4',
        },
      },
      tiles: {},
    }
    await expect(publishEncodedModel(ceramic, encodedModel)).resolves.toMatchSnapshot()
  })

  test('creation flow', async () => {
    const manager = new ModelManager(ceramic)

    const noteSchemaID = await manager.createSchema('Note', {
      $schema: 'http://json-schema.org/draft-07/schema#',
      title: 'Note',
      type: 'object',
      properties: {
        date: {
          type: 'string',
          format: 'date-time',
          title: 'date',
          maxLength: 30,
        },
        text: {
          type: 'string',
          title: 'text',
          maxLength: 4000,
        },
      },
      required: [],
    } as any)
    const notesListSchemaID = await manager.createSchema('NotesList', {
      $schema: 'http://json-schema.org/draft-07/schema#',
      title: 'NotesList',
      type: 'object',
      properties: {
        notes: {
          type: 'array',
          title: 'notes',
          items: {
            type: 'object',
            title: 'NoteItem',
            properties: {
              id: {
                $comment: `cip88:ref:${manager.getSchemaURL(noteSchemaID) as string}`,
                type: 'string',
              },
              title: {
                type: 'string',
                title: 'title',
                maxLength: 100,
              },
            },
            required: [],
          },
        },
      },
      required: [],
    } as any)
    expect(manager.schemas).toEqual(['Note', 'NotesList'])

    await manager.createDefinition('myNotes', {
      name: 'notes',
      description: 'My notes',
      schema: manager.getSchemaURL(notesListSchemaID) as string,
    })
    expect(manager.schemas).toEqual(['DataStoreDefinition', 'DataStoreIndex', 'Note', 'NotesList'])
    expect(manager.definitions).toEqual(['myNotes'])

    expect(manager.toJSON()).toBeDefined()
  })
})
