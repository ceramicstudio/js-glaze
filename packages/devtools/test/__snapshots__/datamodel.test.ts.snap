// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`datamodel publish encoded model 1`] = `
Object {
  "definitions": Object {
    "myNotes": "kjzl6cwe1jw145d00g4tuhpps6ozv30hb30cnlaevxmhy8hqbo2x8w58bupuhua",
  },
  "schemas": Object {
    "DataStoreDefinition": "ceramic://k3y52l7qbv1fry1fp4s0nwdarh0vahusarpposgevy0pemiykymd2ord6swtharcw",
    "DataStoreIndex": "ceramic://k3y52l7qbv1fryjn62sggjh1lpn11c56qfofzmty190d62hwk1cal1c7qc5he54ow",
    "Note": "ceramic://k3y52l7qbv1frxr40y3mwy5qlywfty4sfcqrmv4zckjcdnylk5jdq7cgq66r0acjk",
    "NotesList": "ceramic://k3y52l7qbv1fryc2v942v5n6gzesjlg0zu49gnufb97ohhmeym56dme5dit1dmfwg",
  },
  "tiles": Object {},
}
`;
